### DISCUSSION TITLE
Explicit Targeting and Unknown Packages

### DISCUSSION THREAD

**Discussion by MrAbidPatel** [microsoft/winget-cli#3066] [Q&A] [answered]
I'm trying to use `winget upgrade --all -h` to update software on my Windows 11 PC, but I keep getting the following message:

```
The following packages have an upgrade available, but require explicit targeting for upgrade
Name                         Id              Version     Available  Source
--------------------------------------------------------------------------
Teams Machine-Wide Installer Microsoft.Teams 1.5.0.30767 1.6.00.376 winget
8 package(s) have version numbers that cannot be determined. Use --include-unknown to see all results.
1 package(s) is pinned and needs to be explicitly upgraded.
```

I've tried using `winget upgrade --all -h --include-unknown` but that refuses to work:

```
No installed package found matching input criteria.

The following packages have an upgrade available, but require explicit targeting for upgrade:
Name                         Id              Version     Available  Source
--------------------------------------------------------------------------
Teams Machine-Wide Installer Microsoft.Teams 1.5.0.30767 1.6.00.376 winget
1 package(s) is pinned and needs to be explicitly upgraded.
```

Any ideas how I can upgrade the 8 packages and also how I can upgrade the package that needs explicit targeting?

**Comment by Trenly** [comment, marked as answer]
Explicit Targeting means that you need to pass the ID to upgrade it and it will be excluded from `--all`. For Teams you would need to do `winget upgrade Microsoft.Teams`

Regarding the unknown versions, it's likely they are already up to date. WinGet uses the version in the registry to determine if an application is up to date or not. If the installer doesn't write the version to the registry, it will always show as unknown. So, even after an upgrade, it's possible (and even likely) that the version will remain unknown. The other part of it is that a newer version may not be *applicable*, which means that even though it is *available*, there is something preventing it from being installed; but, these messages aren't always shown when using `--all`. If you run `winget upgrade --include-unknown` it should list out the items it detects as needing upgrade. If any show up with version unknown, try specifying the ID and see if you get a specific error message

**Reply by github-account1111** [reply]
> you need to pass the ID to upgrade it and it will be excluded from `--all`

Why is that needed? Seems very arbitrary.

Is it because I installed the package like so? `winget install --id DigitalScholar.Zotero --interactive --scope machine`

**Comment by MrAbidPatel** [comment]
I tried running  `winget upgrade Microsoft.Teams` and I got the following message: `No installed package found matching input criteria.`

However, when I then try `winget upgrade --all`, I get the following message again:

```
No installed package found matching input criteria.

The following packages have an upgrade available, but require explicit targeting for upgrade:
Name                         Id              Version     Available  Source
--------------------------------------------------------------------------
Teams Machine-Wide Installer Microsoft.Teams 1.5.0.30767 1.6.00.376 winget
8 package(s) have version numbers that cannot be determined. Use --include-unknown to see all results.
1 package(s) is pinned and needs to be explicitly upgraded.
```

Running Windows 11 Pro (22H2) and running winget using Terminal (Run as Administrator)

**Reply by Trenly** [reply]
Looks like this is a Matching error where the `Teams Machine-Wide Installer` is being matched incorrectly for `winget upgrade --all` but the entry for `Microsoft Teams` is being matched correctly for `winget upgrade Microsoft.Teams`.  There are several known issues around matching, I'll tag @denelon to see if we have one capturing this specific behavior or not

**Reply by denelon** [reply]
This is the latest report:
* https://github.com/microsoft/winget-cli/issues/3063

It will likely get closed as a duplicate, but you will still be able to track it from that issue.

**Comment by chrisxfire** [comment]
Same issue; same symptoms.  Let me know if I can provide some logs to help.

>OS Name:                   Microsoft Windows 10 Pro
>OS Version:                10.0.19045 N/A Build 19045

**Comment by denelon** [comment]
Unfortunately we haven't found a good solution for Teams. What appears to be happening is when Team is installed for the current "user" rather than "machine" another package is installed "Teams Machine-Wide Installer". The matching logic detects that as the "first" match encountered. We're working on several issues related to matching and we're adding the "Area-Matching" label to those. We've started to work on the first few aspects of this problem. We will continue to refine the logic as we encounter sub-classes of this problem.

**Reply by GeekyBryan** [reply]
Same on Win11 22H2.  No matter what I do it does not work.  Trying by ID gives:
     No installed package found matching input criteria.
     NativeCommandExitException: Program "winget.exe" ended with non-zero exit code: -1978335212.

Same by name, and also using --version.  Broken logic somewhere.

**Comment by darrellarbaugh** [comment]
--name 'Microsoft Teams or --name 'Teams Machine-Wide Installer' works just fine.  However, thre is a collision with Team installs when both items are installed (likely due to a managed application installation vs winget install).

**Reply by marc3lom** [reply]
I ran winget remove --name 'Teams Machine-Wide Installer' and then winget install Microsoft.Teams and now it doesn't appear as a pinned package when I run winget upgrade.

**Reply by gitnol** [reply]
That solved it for me, too.

**Reply by zvekan** [reply]
me too

**Comment by Inetgate** [comment]
I am experiencing similar issue on Win 11 23H2 until today.
For example, I run "winget upgrade", I saw as below:

> winget upgrade
次のパッケージにはアップグレードを適用可能ですが、アップグレードには明示的対象化が必要です:
名前                    ID                      バージョン   利用可能     ソース
--------------------------------------------------------------------------------
Microsoft Teams classic Microsoft.Teams.Classic 1.6.00.29964 1.7.00.19353 winget
Obsidian                Obsidian.Obsidian       1.5.8        1.6.7        winget


Today, at first I run "winget upgrade Obsidian.Obsidian" and completed. 
Then I run "winget upgrade", those messages are disappeared.

So, I wonder that remaining Microsoft.Teams.Classic is caused by another issue.
I mean if some another package is remaining, Microsoft.Teams.Classic is remaining.

Regards.
